src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -546,33 +546,73 @@ private void doScheduledSearchForRound (RaptorState outputState) {
-                        //  A----B
-                        //       |
-                        //       O
-                        // In this setup, the earliestBoardTime at A is less than (before) the earliestBoardTime at B.
-                        // It might be possible to board a certain trip at B, but only a later trip at A. Because A
-                        // is evaluated first (providing updates to downstream stops), we need to check for earlier
-                        // trips that could be boarded at B.
-                        int bestTripIdx = onTrip;
-                        while (--bestTripIdx >= 0) {
-                            TripSchedule trip = pattern.tripSchedules.get(bestTripIdx);
-                            if (trip.headwaySeconds != null || !servicesActive.get(trip.serviceCode)) {
+
+                        // We have already boarded a trip on this pattern at an upstream stop; check for cases in which
+                        // it would be preferable to board the pattern at this stop instead.
+
+                        TripSchedule onTripSchedule = pattern.tripSchedules.get(onTrip);
+
+                        // First, check whether boarding the same trip at this stop instead of the upstream one would
+                        // be preferred. This will not affect total travel time, but it will affect the breakdown of
+                        // walk vs. wait. So if needed as an optimization, these blocks could be disabled unless
+                        // request.computePaths = true;
+                        boolean boardHereInstead = false;
+                        if (inputState.previous == null) {
+                            // In the first transit round, arriving at this stop before arriving at the upstream stop
+                            // implies shorter access time. And arriving earlier or at the same time implies more
+                            // buffer time to board the trip, so boarding at this stop is preferable to boarding at
+                            // the upstream stop when the non-strict inequality is satisfied.
+                            if (inputState.bestTimes[stop] <= inputState.bestTimes[boardStop]) {
+                                boardHereInstead = true;
+                            }
+                        } else {
+                            // In subsequent transit rounds, we want to compare the length of the transfer legs.
+                            // Boarding a trip at this stop is preferable if it requires a shorter transfer than
+                            // boarding the same trip at the upstream stop.
+                            int upstreamTransferFromStop = inputState.transferStop[boardStop];
+                            if (upstreamTransferFromStop != -1) {
+                                int upstreamTransferTime =
+                                        inputState.bestTimes[boardStop] - inputState.previous.bestTimes[upstreamTransferFromStop];
```

Comment:
If I'm not mistaken, we don't need to look at the previous round at all. We can look at the bestNonTransferTime for the upstreamTransferFromStop in the input state. The transferStop tells us which stop we transferred from within that same round, if any. The difference between those two arrival times should be proportionate to the the transfer distance.

If there is no need to look at the previous round, then I'm hopeful that these two main clauses (`if (inputState.previous == null) ... else`)  can somehow be collapsed into one. It seems to me like there should be a single expression that works to break ties in all rounds.

Reply:
The tie breaking rule you’re using in the latest updates in 782936e would do the job. I was just wondering if there was a single expression with less conditionals that would achieve something similar. One alternative might be the one mentioned in another comment below: maximizing wait time instead of minimizing walk time (which should be equivalent when boarding the same trip at different stops). I'll provide some more detail on that comment.